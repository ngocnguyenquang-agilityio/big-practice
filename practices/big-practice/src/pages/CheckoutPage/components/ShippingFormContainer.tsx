// Libs
import { Link } from 'react-router-dom';
import { useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import * as yup from 'yup';

// Components
import { Button } from '@components/Button/Button';
import { Icon } from '@components/Icon/Icon';
import { Input } from '@components/Input/Input';

// Icons
import logoIcon from '@assets/logoIcon.svg';

// Constants
import { APP_ROUTERS } from '@constants';

interface IShippingFormSubmit {
  phone: number;
  email: string;
  firstName: string;
  lastName: string;
  address: string;
  city: string;
}

const schema = yup
  .object({
    phone: yup.number().required(),
    email: yup.string().email('Invalid email').required('Email is required'),
    firstName: yup.string().required('First name is required'),
    lastName: yup.string().required('Last name is required'),
    address: yup.string().required('Address is required'),
    city: yup.string().required('City is required'),
  })
  .required();

export const ShippingForm = () => {
  const {
    register,
    formState: { errors },
    handleSubmit,
  } = useForm<IShippingFormSubmit>({
    mode: 'onSubmit',
    defaultValues: {},
    reValidateMode: 'onChange',
    resolver: yupResolver(schema),
  });

  const onSubmit = (data: IShippingFormSubmit) => {
    // TODO: Handle Submit form
    alert(JSON.stringify(data));
  };

  return (
    <>
      <div className='flex items-center justify-center border border-neutral-700 bg-black h-[50px] w-[50px] rounded-xl'>
        <Link
          to={APP_ROUTERS.HOMEPAGE}
          reloadDocument
        >
          <Icon
            svg={logoIcon}
            name='logo-icon'
            width={30}
            height={30}
          />
        </Link>
      </div>
      <form onSubmit={handleSubmit(onSubmit)}>
        <fieldset className='mt-6'>
          <legend className='text-lg font-bold mb-2'>Contact</legend>
          <label
            htmlFor='phone'
            className='block mb-2 text-sm font-medium text-neutral-400'
          >
            Phone
          </label>
          <Input
            type='text'
            id='phone'
            placeholder='Mobile phone number'
            variant={errors?.phone ? 'error' : 'secondary'}
            className='py-3'
            {...register('phone')}
          />
          {errors?.phone?.message && <p className='text-sm text-red-500'>Enter phone number</p>}
        </fieldset>

        <fieldset className='mt-6'>
          <legend className='text-lg font-bold mb-2'>Shipping address</legend>
          <div className='mb-4'>
            <label
              htmlFor='email'
              className='block mb-2 text-sm font-medium text-neutral-400'
            >
              Your email
            </label>
            <Input
              type='text'
              id='email'
              placeholder='name@flowbite.com'
              variant={errors?.email ? 'error' : 'secondary'}
              className='py-3'
              {...register('email')}
            />
            {errors?.email?.message && <p className='text-sm text-red-500'>{errors?.email?.message}</p>}
          </div>
          <div className='grid md:grid-cols-2 md:gap-6'>
            <div className='relative z-0 w-full mb-4 group'>
              <label
                htmlFor='firstName'
                className='block mb-2 text-sm font-medium text-neutral-400'
              >
                First name
              </label>
              <Input
                type='text'
                id='firstName'
                placeholder='First name'
                variant={errors?.email ? 'error' : 'secondary'}
                className='py-3'
                {...register('firstName')}
              />
              {errors?.firstName?.message && <p className='text-sm text-red-500'>{errors?.firstName?.message}</p>}
            </div>
            <div className='relative z-0 w-full mb-4 group'>
              <label
                htmlFor='last-name'
                className='block mb-2 text-sm font-medium text-neutral-400'
              >
                Last name
              </label>
              <Input
                type='text'
                id='last-name'
                placeholder='Last name'
                variant={errors?.lastName ? 'error' : 'secondary'}
                className='py-3'
                {...register('lastName')}
              />
              {errors?.lastName?.message && <p className='text-sm text-red-500'>{errors?.lastName?.message}</p>}
            </div>
          </div>
          <div className='mb-6'>
            <label
              htmlFor='address'
              className='block mb-2 text-sm font-medium text-neutral-400'
            >
              Address
            </label>
            <Input
              type='text'
              id='address'
              placeholder='Address'
              variant={errors?.address ? 'error' : 'secondary'}
              className='py-3'
              {...register('address')}
            />
            {errors?.address?.message && <p className='text-sm text-red-500'>{errors?.address?.message}</p>}
          </div>
          <div className='mb-6'>
            <label
              htmlFor='apartment'
              className='block mb-2 text-sm font-medium text-neutral-400'
            >
              Apartment
            </label>
            <Input
              type='text'
              id='apartment'
              name='apartment'
              placeholder='Apartment, suite, etc.'
              variant='secondary'
              className='py-3'
            />
          </div>
          <div className='mb-6'>
            <label
              htmlFor='city'
              className='block mb-2 text-sm font-medium text-neutral-400'
            >
              City
            </label>
            <Input
              type='text'
              id='city'
              placeholder='City'
              variant={errors?.city ? 'error' : 'secondary'}
              className='py-3'
              {...register('city')}
            />
            {errors?.city?.message && <p className='text-sm text-red-500'>{errors?.city?.message}</p>}
          </div>
          <div className='flex justify-end'>
            <Button
              type='submit'
              className='px-5 py-8 rounded-md'
              size='lg'
            >
              Continue to shipping
            </Button>
          </div>
        </fieldset>
      </form>
    </>
  );
};

const ShippingFormContainer = () => {
  return <ShippingForm />;
};

export default ShippingFormContainer;
