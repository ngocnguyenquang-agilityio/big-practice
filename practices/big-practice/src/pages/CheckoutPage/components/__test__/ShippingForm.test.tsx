// Libs
import { MemoryRouter } from 'react-router-dom';
import { fireEvent, render, RenderResult } from '@testing-library/react';

// Components
import { ShippingForm } from '../ShippingFormContainer';

describe('ShippingForm component', () => {
  let comp: RenderResult;
  beforeEach(() => {
    comp = render(
      <MemoryRouter>
        <ShippingForm />
      </MemoryRouter>,
    );
  });

  test('Should render correctly', () => {
    expect(comp).toBeTruthy();
    expect(comp.getByPlaceholderText('Phone number')).toBeTruthy();
    expect(comp.getByPlaceholderText('name@flowbite.com')).toBeTruthy();
    expect(comp.getByPlaceholderText('First name')).toBeTruthy();
    expect(comp.getByPlaceholderText('Last name')).toBeTruthy();
    expect(comp.getByPlaceholderText('Address')).toBeTruthy();
    expect(comp.getByPlaceholderText('Apartment')).toBeTruthy();
    expect(comp.getByPlaceholderText('City')).toBeTruthy();
    expect(comp.getByText(/continue to shipping/i, { selector: 'button' })).toBeTruthy();
  });

  test('Should match snapshot', () => {
    expect(comp).toMatchSnapshot();
  });

  test('Should render error messages', () => {
    const button = comp.getByText(/continue to shipping/i, { selector: 'button' });
    fireEvent.click(button);

    expect(comp.findByText('Phone is required')).toBeTruthy();
    expect(comp.findByText('Email is required')).toBeTruthy();
    expect(comp.findByText('First name is required')).toBeTruthy();
    expect(comp.findByText('Last name is required')).toBeTruthy();
    expect(comp.findByText('Address is required')).toBeTruthy();
    expect(comp.findByText('City is required')).toBeTruthy();
  });
});

describe('ShippingForm component phone number field', () => {
  let comp: RenderResult;
  beforeEach(() => {
    comp = render(
      <MemoryRouter>
        <ShippingForm />
      </MemoryRouter>,
    );
  });

  test('Should render error message for phone number field', async () => {
    const phoneInput = comp.getByPlaceholderText('Phone number');
    const button = comp.getByText(/continue to shipping/i, { selector: 'button' });
    fireEvent.change(phoneInput, { target: { value: 'this is not phone number' } });
    await fireEvent.click(button);

    expect(comp.findByText('Phone number is not valid')).toBeTruthy();
  });
});

// TODO: Update UT for the rest input field the same as phone number
