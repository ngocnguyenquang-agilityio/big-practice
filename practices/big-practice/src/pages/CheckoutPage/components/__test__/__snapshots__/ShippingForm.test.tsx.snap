// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShippingForm component Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex items-center justify-center border border-neutral-700 bg-black h-[50px] w-[50px] rounded-xl"
      >
        <a
          href="/"
        >
          <img
            alt="logo-icon"
            height="30"
            src=""
            width="30"
          />
        </a>
      </div>
      <form>
        <fieldset
          class="mt-6"
        >
          <legend
            class="text-lg font-bold mb-2"
          >
            Contact
          </legend>
          <label
            class="block mb-2 text-sm font-medium text-neutral-400"
          >
            Phone
          </label>
          <input
            class="w-full rounded-lg border border-neutral-800 bg-transparent px-4 text-sm text-white placeholder:text-neutral-400 focus:outline-none focus:border-blue-600 py-3"
            placeholder="Phone number"
            type="text"
            value=""
          />
        </fieldset>
        <fieldset
          class="mt-6"
        >
          <legend
            class="text-lg font-bold mb-2"
          >
            Shipping address
          </legend>
          <div
            class="mb-4"
          >
            <label
              class="block mb-2 text-sm font-medium text-neutral-400"
            >
              Your email
            </label>
            <input
              class="w-full rounded-lg border border-neutral-800 bg-transparent px-4 text-sm text-white placeholder:text-neutral-400 focus:outline-none focus:border-blue-600 py-3"
              placeholder="name@flowbite.com"
              type="text"
              value=""
            />
          </div>
          <div
            class="grid md:grid-cols-2 md:gap-6"
          >
            <div
              class="relative z-0 w-full mb-4 group"
            >
              <label
                class="block mb-2 text-sm font-medium text-neutral-400"
              >
                First name
              </label>
              <input
                class="w-full rounded-lg border border-neutral-800 bg-transparent px-4 text-sm text-white placeholder:text-neutral-400 focus:outline-none focus:border-blue-600 py-3"
                placeholder="First name"
                type="text"
                value=""
              />
            </div>
            <div
              class="relative z-0 w-full mb-4 group"
            >
              <label
                class="block mb-2 text-sm font-medium text-neutral-400"
              >
                Last name
              </label>
              <input
                class="w-full rounded-lg border border-neutral-800 bg-transparent px-4 text-sm text-white placeholder:text-neutral-400 focus:outline-none focus:border-blue-600 py-3"
                placeholder="Last name"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="mb-6"
          >
            <label
              class="block mb-2 text-sm font-medium text-neutral-400"
            >
              Address
            </label>
            <input
              class="w-full rounded-lg border border-neutral-800 bg-transparent px-4 text-sm text-white placeholder:text-neutral-400 focus:outline-none focus:border-blue-600 py-3"
              placeholder="Address"
              type="text"
              value=""
            />
          </div>
          <div
            class="mb-6"
          >
            <label
              class="block mb-2 text-sm font-medium text-neutral-400"
            >
              Apartment
            </label>
            <input
              class="w-full rounded-lg border border-neutral-800 bg-transparent px-4 text-sm text-white placeholder:text-neutral-400 py-3"
              placeholder="Apartment"
              type="text"
              value=""
            />
          </div>
          <div
            class="mb-6"
          >
            <label
              class="block mb-2 text-sm font-medium text-neutral-400"
            >
              City
            </label>
            <input
              class="w-full rounded-lg border border-neutral-800 bg-transparent px-4 text-sm text-white placeholder:text-neutral-400 focus:outline-none focus:border-blue-600 py-3"
              placeholder="City"
              type="text"
              value=""
            />
          </div>
          <div
            class="flex justify-end"
          >
            <button
              class="inline-flex items-center text-white justify-center text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 ring-offset-slate-950 focus-visible:ring-slate-300 bg-blue-600 hover:opacity-90 disabled:cursor-not-allowed disabled:opacity-50 h-11 px-5 py-8 rounded-md"
              type="submit"
            >
              Continue to shipping
            </button>
          </div>
        </fieldset>
      </form>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex items-center justify-center border border-neutral-700 bg-black h-[50px] w-[50px] rounded-xl"
    >
      <a
        href="/"
      >
        <img
          alt="logo-icon"
          height="30"
          src=""
          width="30"
        />
      </a>
    </div>
    <form>
      <fieldset
        class="mt-6"
      >
        <legend
          class="text-lg font-bold mb-2"
        >
          Contact
        </legend>
        <label
          class="block mb-2 text-sm font-medium text-neutral-400"
        >
          Phone
        </label>
        <input
          class="w-full rounded-lg border border-neutral-800 bg-transparent px-4 text-sm text-white placeholder:text-neutral-400 focus:outline-none focus:border-blue-600 py-3"
          placeholder="Phone number"
          type="text"
          value=""
        />
      </fieldset>
      <fieldset
        class="mt-6"
      >
        <legend
          class="text-lg font-bold mb-2"
        >
          Shipping address
        </legend>
        <div
          class="mb-4"
        >
          <label
            class="block mb-2 text-sm font-medium text-neutral-400"
          >
            Your email
          </label>
          <input
            class="w-full rounded-lg border border-neutral-800 bg-transparent px-4 text-sm text-white placeholder:text-neutral-400 focus:outline-none focus:border-blue-600 py-3"
            placeholder="name@flowbite.com"
            type="text"
            value=""
          />
        </div>
        <div
          class="grid md:grid-cols-2 md:gap-6"
        >
          <div
            class="relative z-0 w-full mb-4 group"
          >
            <label
              class="block mb-2 text-sm font-medium text-neutral-400"
            >
              First name
            </label>
            <input
              class="w-full rounded-lg border border-neutral-800 bg-transparent px-4 text-sm text-white placeholder:text-neutral-400 focus:outline-none focus:border-blue-600 py-3"
              placeholder="First name"
              type="text"
              value=""
            />
          </div>
          <div
            class="relative z-0 w-full mb-4 group"
          >
            <label
              class="block mb-2 text-sm font-medium text-neutral-400"
            >
              Last name
            </label>
            <input
              class="w-full rounded-lg border border-neutral-800 bg-transparent px-4 text-sm text-white placeholder:text-neutral-400 focus:outline-none focus:border-blue-600 py-3"
              placeholder="Last name"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="mb-6"
        >
          <label
            class="block mb-2 text-sm font-medium text-neutral-400"
          >
            Address
          </label>
          <input
            class="w-full rounded-lg border border-neutral-800 bg-transparent px-4 text-sm text-white placeholder:text-neutral-400 focus:outline-none focus:border-blue-600 py-3"
            placeholder="Address"
            type="text"
            value=""
          />
        </div>
        <div
          class="mb-6"
        >
          <label
            class="block mb-2 text-sm font-medium text-neutral-400"
          >
            Apartment
          </label>
          <input
            class="w-full rounded-lg border border-neutral-800 bg-transparent px-4 text-sm text-white placeholder:text-neutral-400 py-3"
            placeholder="Apartment"
            type="text"
            value=""
          />
        </div>
        <div
          class="mb-6"
        >
          <label
            class="block mb-2 text-sm font-medium text-neutral-400"
          >
            City
          </label>
          <input
            class="w-full rounded-lg border border-neutral-800 bg-transparent px-4 text-sm text-white placeholder:text-neutral-400 focus:outline-none focus:border-blue-600 py-3"
            placeholder="City"
            type="text"
            value=""
          />
        </div>
        <div
          class="flex justify-end"
        >
          <button
            class="inline-flex items-center text-white justify-center text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 ring-offset-slate-950 focus-visible:ring-slate-300 bg-blue-600 hover:opacity-90 disabled:cursor-not-allowed disabled:opacity-50 h-11 px-5 py-8 rounded-md"
            type="submit"
          >
            Continue to shipping
          </button>
        </div>
      </fieldset>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
